pkg/controller/machineset/delete_policy.go

```diff
@@ -17,57 +17,116 @@ limitations under the License.
 package machineset
 
 import (
+	"math"
+	"sort"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
 )
 
-type deletePriority int
+type deletePriority float64
 
 const (
-	mustDelete   deletePriority = 100
-	betterDelete deletePriority = 50
-	couldDelete  deletePriority = 20
-	// mustNotDelete deletePriority = 0
+
+	// DeleteNodeAnnotation marks nodes that will be given priority for deletion
+	// when a machineset scales down. This annotation is given top priority on all delete policies.
+	DeleteNodeAnnotation = "cluster.k8s.io/delete-machine"
+
+	mustDelete    deletePriority = 100.0
+	betterDelete  deletePriority = 50.0
+	couldDelete   deletePriority = 20.0
+	mustNotDelete deletePriority = 0.0
+
+	secondsPerTenDays float64 = 864000
 )
 
 type deletePriorityFunc func(machine *v1alpha1.Machine) deletePriority
 
-func simpleDeletePriority(machine *v1alpha1.Machine) deletePriority {
+// maps the creation timestamp onto the 0-100 priority range
+func oldestDeletePriority(machine *v1alpha1.Machine) deletePriority {
+	if machine.DeletionTimestamp != nil && !machine.DeletionTimestamp.IsZero() {
+		return mustDelete
+	}
+	if machine.ObjectMeta.Annotations != nil && machine.ObjectMeta.Annotations[DeleteNodeAnnotation] != "" {
+		return mustDelete
+	}
+	if machine.Status.ErrorReason != nil || machine.Status.ErrorMessage != nil {
+		return mustDelete
+	}
+	if machine.ObjectMeta.CreationTimestamp.Time.IsZero() {
+		return mustNotDelete
+	}
+	d := metav1.Now().Sub(machine.ObjectMeta.CreationTimestamp.Time)
+	if d.Seconds() < 0 {
+		return mustNotDelete
+	}
+	return deletePriority(float64(mustDelete) * (1.0 - math.Exp(-d.Seconds()/secondsPerTenDays)))
```

Comment:
I'm curious - why seconds per 10 days?

Reply:
The function maps [0, Inf) -> [0, 1).  With exponential decay, the parameter puts the half-life at just about a week, i.e. 1 week ~= 0.5 priority, 2 weeks ~= 0.75 priority etc. Arbitrary but reasonable?

Reply:
You mean 0-100, right?

Reply:
scaled by `float64(mustDelete)` to the appropriate range, yeah.